lib/class-wp-annotation-utils.php

```diff
@@ -0,0 +1,820 @@
+					return false;
+				}
+				return true;
+
+			case 'TextQuoteSelector':
+				$allow_keys = array(
+					'type',
+					'exact',
+					'prefix',
+					'suffix',
+				);
+				if ( array_diff_key( $selector, array_fill_keys( $allow_keys, 0 ) ) ) {
+					return false;
+				} elseif ( ! isset( $selector['exact'] ) || ! is_string( $selector['exact'] ) ) {
+					return false;
+				} elseif ( isset( $selector['prefix'] ) && ! is_string( $selector['prefix'] ) ) {
+					return false;
+				} elseif ( isset( $selector['suffix'] ) && ! is_string( $selector['suffix'] ) ) {
+					return false;
+				} elseif ( '' === $selector['exact'] ) {
+					return false;
+				}
+				return true;
+
+			case 'TextPositionSelector':
+			case 'DataPositionSelector':
+				$allow_keys = array(
+					'type',
+					'start',
+					'end',
+				);
+				if ( array_diff_key( $selector, array_fill_keys( $allow_keys, 0 ) ) ) {
+					return false;
+				} elseif ( ! isset( $selector['start'] ) || ! is_int( $selector['start'] ) || 0 > $selector['start'] ) {
+					return false;
+				} elseif ( ! isset( $selector['end'] ) || ! is_int( $selector['end'] ) || 0 > $selector['end'] ) {
+					return false;
+				}
+				return true;
+
+			case 'SvgSelector':
+				/*
+				 * @TODO SVG selectors are disabled for the time being. See {@see
+				 * WP_Annotation_Utils::$allow_selectors} for further details.
+				 *
+				 * Please DO NOT ENABLE until a better security scan can be performed here.
+				 */
+				$allow_keys = array(
+					'type',
+					'id',    // URL leading to an SVG file.
+					'value', // Inline SVG markup.
+				);
+				if ( array_diff_key( $selector, array_fill_keys( $allow_keys, 0 ) ) ) {
+					return false;
+				} elseif ( empty( $selector['id'] ) && empty( $selector['value'] ) ) {
+					return false;
+				} elseif ( ! empty( $selector['id'] ) && ! wp_parse_url( $selector['id'] ) ) {
+					return false;
+				} elseif ( ! empty( $selector['value'] ) && false === stripos( (string) $selector['value'], '</svg>' ) ) {
+					return false;
+				}
+				return true;
+
+			case 'RangeSelector':
+				$allow_keys = array(
+					'type',
+					'startSelector',
+					'endSelector',
+				);
+				if ( array_diff_key( $selector, array_fill_keys( $allow_keys, 0 ) ) ) {
+					return false;
+				} elseif ( empty( $selector['startSelector'] ) || empty( $selector['endSelector'] ) ) {
+					return false;
+				} elseif ( ! self::is_valid_selector( $selector['startSelector'], true ) ) {
+					return false;
+				} elseif ( ! self::is_valid_selector( $selector['endSelector'], true ) ) {
+					return false;
+				}
+				return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks if comment content is empty.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $content Content to check.
+	 *
+	 * @return bool           True if empty, if it's not a string, or the blog's charset
+	 *                        is UTF-8 and it's invalid UTF-8.
+	 */
+	public static function is_content_empty( $content ) {
```

Comment:
One could argue this function is over-the-top. On the other hand, without this level of detail it would be possible to submit something like the example you gave `<span></span>`. Strictly speaking, that's not empty, but it might as well be, because it produces an empty comment.

Another example would be a user who lacks the `unfiltered_html` capability submitting an annotation containing only tags they're not allowed to use, which are stripped away when the content is filtered, leaving an empty string, or empty-ish remnants; e.g, `<span></span>`.

So the point here is to do the best we can at enforcing this rule; i.e., that every annotation must be accompanied by something meaningful, and where `<span></span>` is not considered meaningful, in the same way `''` is not considered meaningful.

Reply:
Can you think of a reason why `<span></span>` should be allowed through server-side?

Reply:
I don't feel that strongly about this. The only thing is that it's pretty complex validation logic for arguably little gain.